File: tests/Passes/Optimization/phase_folding.cpp
/*------------------------------------------------------------------------------
| Part of Tweedledum Project.  This file is distributed under the MIT License.
| See accompanying file /LICENSE for details.
*-----------------------------------------------------------------------------*/
#include "tweedledum/Passes/Optimization/phase_folding.h"

#include "tweedledum/IR/Circuit.h"
#include "tweedledum/IR/Wire.h"
#include "tweedledum/Operators/All.h"

#include "../check_unitary.h"

#include <catch.hpp>

using namespace tweedledum;

TEST_CASE("Trivial phase folding", "[phase_folding][optimization]")
{
    using namespace tweedledum;
    SECTION("Trivial 1 qubit") {
        Circuit circuit;
        Qubit q0 = circuit.create_qubit();
        circuit.apply_operator(Op::T(), {q0});
        circuit.apply_operator(Op::Tdg(), {q0});
        
        Circuit optimized = phase_folding(circuit);
        CHECK(optimized.size() == 0u);
        CHECK(check_unitary(circuit, optimized));
    }
    SECTION("Trivial 2 qubit swap") {
        Circuit circuit;
        Qubit q0 = circuit.create_qubit();
        Qubit q1 = circuit.create_qubit();
        circuit.apply_operator(Op::T(), {q0});
        circuit.apply_operator(Op::Swap(), {q1, q0});
        circuit.apply_operator(Op::Tdg(), {q1});
        
        Circuit optimized = phase_folding(circuit);
        CHECK(optimized.size() == 1u);
        CHECK(check_unitary(circuit, optimized));
    }
}

File: python/tweedledum/qiskit/passes/optimization/phase_folding.py
# -------------------------------------------------------------------------------
# Part of Tweedledum Project.  This file is distributed under the MIT License.
# See accompanying file /LICENSE for details.
# -------------------------------------------------------------------------------
from typing import Union

from qiskit.transpiler.basepasses import TransformationPass
from qiskit.dagcircuit import DAGCircuit

from tweedledum.qiskit import from_qiskit, to_qiskit
from tweedledum.ir import Circuit
from tweedledum.passes import phase_folding


class PhaseFolding(TransformationPass):
    """TODO"""

    def __init__(self):
        """PhaseFolding initializer."""
        super().__init__()

    def run(self, dag: Union[DAGCircuit, Circuit]) -> Union[DAGCircuit, Circuit]:
        """Run the PhaseFolding pass on a circuit.

        Args:
            dag (Union[DAGCircuit, Circuit]): A circuit to map.

        Returns:
            Union[DAGCircuit, Circuit]: An optimized circuit.
        """
        if isinstance(dag, DAGCircuit):
            circuit = from_qiskit(dag)
        else:
            circuit = dag

        opt_circuit = phase_folding(circuit)

        if isinstance(dag, DAGCircuit):
            return to_qiskit(opt_circuit)
        return opt_circuit
